File: Problems/Strings/Longest-Repeating-Character-Replacement/Longest-Repeating-Character-Replacement.cs
using System;
using System.Collections.Generic;

// Using sliding window technique - Time: O(n)
public class Solution
{
    public int CharacterReplacement(string s, int k)
    {
        int i = 0;
        int j = 0;
        int[] cnt = new int[26];
        int n = s.Length;
        
        while (j < n)
        {
            cnt[s[j] - 'A']++;
            j++;
            
            if (j - i - maxElement(cnt, 26) > k)
            {
                cnt[s[i] - 'A']--;
                i++;
            }
        }
        
        return j - i;
    }
    private int maxElement(int[] arr, int length)
    {
        int max = arr[0];
        for (int i = 1; i < length; i++)
        {
            if (arr[i] > max)
            {
                max = arr[i];
            }
        }
        return max;
    }
}

class Program
{
    static void Main(string[] args)
    {
        string s = "ABAB";
        int k = 2;

        Console.WriteLine("Input: s = " + s + ", k = " + k);

        Solution sol = new Solution();
        int result = sol.CharacterReplacement(s, k);

        Console.WriteLine("Output: " + result);
    }
}

File: Problems/Strings/Longest-Repeating-Character-Replacement/Longest-Repeating-Character-Replacement.cpp
#include <iostream>
#include <string>
#include <algorithm>

// Using sliding window technique - Time: O(n)

class Solution {
public:
    int characterReplacement(std::string s, int k)
    {
        int i = 0;
        int j = 0;
        int cnt[26] = {};
        int n = s.size();

        while (j < n)
        {
            cnt[s[j++] - 'A']++;
            if (j - i - *std::max_element(cnt, cnt + 26) > k)
                cnt[s[i++] - 'A']--;
        }
        return j - i;
    }
};

int main()
{
    std::string s = "ABAB";
    int k = 2;

    std::cout << "Input: s = " << s << ", k = " << k << std::endl;

    Solution sol;
    int result = sol.characterReplacement(s, k);

    std::cout << "Output: " << result << std::endl;

    return 0;
}